app/models/comment.rb

```diff
@@ -98,9 +113,22 @@ class Comment < AbstractModel
   after_create :notify_users
   after_create :oil_and_water
 
+  scope :by_user,
+        ->(user) { where(user: user) }
+  scope :for_user,
```

Comment:
I like that you've done this without a join.  But do we need to send the list of target ids back to the webserver at all?  Can we do this as a subquery instead?  Or is this somehow being done as a subquery under the hood?  Actually I really wonder how this _is_ being done under the hood as it is!

I think this would be easier to do if we did the polymorphism the way rss_logs does it: select where observation_id in (select id from observations where user_id = 1234) or name_id in (...) or ...

But something somewhat equivalent should be possible with comments: select where target_type = "Observation" and target_id in (select ...) or target_type = "Name" and target_id in (select ...) or etc.

Can that be built in a for loop somehow?  Same number of queries on the db, ultimately, but all of the subqueries' results stay local on the db.  Is that any better?

Reply:
I think that's exactly what it does, right? Ohhh, i get it, you mean bake the loop into the sql so the db does it, not in ruby.

`where(target: proc)` or something?  

I don't know how to do that. I could chain the `where().or(where())` scopes maybe by hardcoding each model. The current scope is nice because the target types are pulled off the class definition further down, so if the target classes change... they only have to change in one place.

Let's set that as an aspiration and i'll put a note in the comments. This scope isn't called anywhere yet, it's just another "note to self", for when we try to summit the Query class with scopes.

Come to think of it, it should probably `pluck` the id, right?

Reply:
Aspiration, yes!  The pluck is a good idea to do immediately.

Just messin' around, can you do this?

`
Comment.all_types do |type|
  model = get_model_from_type_tag(type)
  scope = scope.or.where(target_type: type, target_id: model.where(user: user))
end
`

Reply:
I'm intrigued, because i don't get how chaining works in scopes yet.

Consider this attempt, class method defs included for clarity. This doesn't fly (nor with substituting `self` for `scope`). The first problem is we have to start with a `where`, not an `or`, or else the initial scope will be `Comment.all` and then we're just `or`-ing onto everything. I'm doing this in the rails console and inspecting results.

```ruby
  scope :for_user,
        lambda { |user|
          all_types.each do |model|
            scope = scope.or(where(target_type: model.name.underscore.to_sym,
                                   target_id: model.where(user: user)))
          end
        }

  # Returns Array of all models (Classes) which take comments.
  def self.all_types
    all_type_tags.map { |t| t.to_s.camelize.constantize }
  end

  # Returns Array of all valid +target_type+ values (Symbol's).
  def self.all_type_tags
    [:location, :name, :observation, :project, :species_list]
  end
```

Reply:
Ok. Consider this, which works as of Rails 6. 
```ruby
# in the Person model:
scope :john, -> { where(name: "John") }
scope :smith, -> { where(lastname: "Smith") }
scope :john_or_smith, -> { john.or(self.smith) }
```
Can we iterate to make a bunch of scopes, and put them together?

Reply:
We can also do this. It's confounding me how to combine them with the above:
```ruby
  [:small, :medium, :large].each do |s|
    scope s, where(size: s) 
  end
``

Reply:
What about something like this?

```
scope :for_user,
  lambda do |user|
    all_types.inject(nil) do |scope, model|
      scope2 = where(target_type: model.name.underscore.to_sym,
                     target_id: model.where(user: user))
      scope ? scope.or(scope2) : scope2
    end
  end
```

Reply:
Less certain this will work, but it would be shorter...

```
scope :for_user,
  lambda do |user|
    all_types.inject do |scope, model|
      where(target_type: model.name.underscore.to_sym,
            target_id: model.where(user: user)).or(scope)
    end
  end
```

Reply:
The first one works. What the heck? No comprende.

I'm guessing `nil` pushes nil as the first el of the array. 
How are you iterating two args? What is `do |scope, model|`? in other words

Reply:
Look up how inject works.  I hate that method, it always hurts my brain.  It is equivalent to

```
lambda do |user|
  scope = nil
  all_types.each do |model|
    scope2 =where(blah)
    scope = scope ? scope.or(scope2) : scope2
  end
  scope
end
```

Reply:
Ugh. Gross. But impressive. Kind of the Cordyceps of iterators.

Reply:
Got to respect it.

Reply:
I'm so impressed.
```ruby
3.1.2 :002 > Comment.for_user(252).to_sql
 => "SELECT `comments`.* FROM `comments` WHERE (((`comments`.`target_type` = 'location') AND (`comments`.`target_id` IN (SELECT `locations`.`id` FROM `locations` WHERE `locations`.`user_id` = 252))) OR ((`comments`.`target_type` = 'name') AND (`comments`.`target_id` IN (SELECT `names`.`id` FROM `names` WHERE `names`.`user_id` = 252))) OR ((`comments`.`target_type` = 'observation') AND (`comments`.`target_id` IN (SELECT `observations`.`id` FROM `observations` WHERE `observations`.`user_id` = 252))) OR ((`comments`.`target_type` = 'project') AND (`comments`.`target_id` IN (SELECT `projects`.`id` FROM `projects` WHERE `projects`.`user_id` = 252))) OR ((`comments`.`target_type` = 'species_list') AND (`comments`.`target_id` IN (SELECT `species_lists`.`id` FROM `species_lists` WHERE `species_lists`.`user_id` = 252))))"                                                                          
```

Reply:
Woah!!  Check it out!  It works!  What a crazy query!  I love it!  Those sorts of puzzles are fun... when we can actually work out the solution(!)

Reply:
Wait a minute. Are we doing an unnecessary check for `target_type`? It's not necessary to specify the `comment` record.

Couldn't we just have it scanning the `target_id` column only? Or is this faster.